crates/uv-auth/src/middleware.rs

```diff
@@ -69,14 +72,39 @@ impl Default for AuthMiddleware {
 
 #[async_trait::async_trait]
 impl Middleware for AuthMiddleware {
+    /// Handle authentication for a request.
+    ///
+    /// If the request has a username and password
+    ///
+    /// - Perform the request
+    /// - Add the username and password to the cache if successful
+    ///
+    /// If the request only has a username:
+    ///
+    /// - Check the cache (realm key) for a password
+    /// - Check the netrc for a password
+    /// - Check the keyring for a password
+    /// - Perform the request
+    /// - Add the username and password to the cache if successful
+    ///
+    /// If the request has no authentication:
+    ///
+    /// - Check the cache (url key)
+    /// - Perform the request
+    /// - On 401 or 403, check for authentication
+    ///     - Check the cache (realm key) for the username and password
+    ///     - Check the netrc for a username and password
+    ///     - Perform the request again if found
+    ///     - Add the username and password to the cache if successful
```

Comment:
Why do we perform the request twice?

Reply:
I believe this matches pip's behavior. We can't _know_ if we're supposed to send authentication for a URL unless it fails. For example, if you try to install two GitHub projects one private and one public, we can attach authentication from the private project to the public project which will fail.

See https://github.com/astral-sh/uv/issues/3123#issuecomment-2064706021

Reply:
Wouldn't the prefix-based URL matching avoid us failing on the two-GitHubs case though?

Reply:
In the "two GitHubs" case, what is an example of the two URLs and the credentials that would be attached to them by the user? Like, the URL for the private project won't be a prefix of the public project, right? So we wouldn't try to reuse the credentials anyway...?

Reply:
So.. if we make an authenticated request we will not perform a retry. The second request is only made if we failed to find cached credentials for the prefix. This ensures that we only make a request with cached realm-level credentials if necessary. 

Let's say we see `github.com/private`, we'll cache credentials at the realm and URL-level for this.
Subsequent requests to `github.com/private/foo` will use the cached URL-level credentials, yay we avoided sending a bad request to realize we needed credentials.
If a subsequent request to `github.com/private/bar` fails, we won't perform a second request.

Now if we see `github.com/other-private`, we'll have a URL cache miss. We don't know it's private so we attempt a request. This request fails so we try the cached credentials for the realm — these work yay.

Now we see `github.com/public`. We'll have a URL cache miss but the request just succeeds — yay.